Implement T::Enum#=== intrinsic

aprocter-stripe opened the pull request:
The root of #4115 seems to be that the `case` desugars to `===` on the singletons for the enums, not the classes. Implementing an intrinsic for `T::Enum#===` seems to me like a reasonable way to fix it, ~~but I'm tossing a wip PR up first to get feedback on whether it's the right approach~~.

### Motivation
Fixes #4115

### Test plan
See included automated tests.

aprocter-stripe commented:
(Just for some additional context: this does seem to fix the errors in #4115, and I think it matches the behavior of the runtime [see https://github.com/sorbet/sorbet/blob/b12cbff755c13f2c659ccfa4688e78729d8faa46/gems/sorbet-runtime/lib/types/enum.rb#L220-L233] except for the `legacy_t_enum_migration_mode`.)

aprocter-stripe commented on core/types/calls.cc:
```diff
@@ -3010,6 +3010,31 @@ class Module_tripleEq : public IntrinsicMethod {
     }
 } Module_tripleEq;
 
+class T_Enum_tripleEq : public IntrinsicMethod {
```
Note that the implementation is based on that of `Module_tripleEq` above.

elliottt replied:
Oh, `Module_tripleEq` wasn't firing because the receiver was a value of one of the generated enum classes?

aprocter-stripe replied:
Yup!

elliottt reviewed (changes requested):
I think that this approach makes sense, given that the receiver of `===` before was one of the singleton enum values instead of a class/module. What do you think @jez?

jez commented on core/types/calls.cc:
```diff
@@ -3010,6 +3010,31 @@ class Module_tripleEq : public IntrinsicMethod {
+            return;
+        }
+        if (Types::glb(gs, rhs, lhs).isBottom()) {
+            res.returnType = Types::falseClass();
```
Can you make sure to write a test for this?

I think that this should work:

```ruby
T.reveal_type(MyEnum::Y.===(MyEnum::X))
```

also it would be great to have both the examples from #4115 committed as test cases.

jez commented:
> I think that this approach makes sense, given that the receiver of `===` before was one of the singleton enum values instead of a class/module. What do you think @jez?

This is the best approach. Thanks!

elliottt reviewed (approved):
Nice, great tests!

aprocter-stripe commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqWM7IQSVLNGVZ
→ https://go/builds/bui_JqWNbeFJ6YHUR4

aprocter-stripe commented:
Looks like I goofed something up here.

```ruby
# typed: true
extend T::Sig

class MyEnum < T::Enum ; end

sig {params(x: T::Enum, y: MyEnum).void}
def f(x, y)
  T.reveal_type(x===y)
end

sig {params(x: MyEnum, y: MyEnum).void}
def g(x, y)
  T.reveal_type(x===y)
end
```

On sorbet.run, both of the above reveal as `T::Boolean`. On this branch they reveal as `T::TrueType`, which is uh wrong. :) Will look into what I did wrong in cargo-culting from `Module#===`.

aprocter-stripe commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqcojtMIA1LKLN
→ https://go/builds/bui_JqcoNstf71fgLx

aprocter-stripe commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqdJt09NHtwwq7
→ https://go/builds/bui_JqdJ8CzkfQ6QI5

aprocter-stripe commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqfS5Hym0W3K8i
→ https://go/builds/bui_JqfSnnD4hB2MoH

aprocter-stripe commented:
So I fixed a couple things:

1. Unbroke behavior when `===`-comparing abstract enum things (i.e. `x===y` where x and y have `T::Enum` or `SomeEnum`; initially the issues here didn't reveal themselves since I was just looking at desugared `case` statements).
2. Added support for `String` on the rhs, to support `legacy_t_enum_migration_mode` (just returns `T::Boolean`).

I'm feeling reasonably good about the correctness here, and I have a green build on Stripe's codebase that doesn't require anything surprising, but since the logic has changed quite a bit a second review is probably in order.

elliottt commented on core/types/calls.cc:
```diff
@@ -3072,6 +3072,59 @@ class Module_tripleEq : public IntrinsicMethod {
+    return superClass.exists() && superClass == Symbols::T_Enum();
+}
+
+// Returns true if the referenced class is a grandchild of T::Enum, i.e., one of the singleton classes for the enum
+// values.
+static bool isEnumValue(const GlobalState &gs, ClassOrModuleRef klass) {
```
```suggestion
// Returns true if the referenced class is a grandchild of T::Enum, i.e., one of the classes for the singleton enum
// values.
static bool isEnumValueClass(const GlobalState &gs, ClassOrModuleRef klass) {
```

elliottt commented on core/types/calls.cc:
```diff
@@ -3072,6 +3072,59 @@ class Module_tripleEq : public IntrinsicMethod {
+        bool must_exist = false;
+        auto lhs_unwrapped = unwrapSymbol(gs, lhs, must_exist);
+        auto rhs_unwrapped = unwrapSymbol(gs, rhs, must_exist);
+        if (Types::isSubType(gs, rhs, lhs) && isEnumValue(gs, lhs_unwrapped) && isEnumValue(gs, rhs_unwrapped)) {
```
```suggestion
        if (Types::isSubType(gs, rhs, lhs) && isEnumValueClass(gs, lhs_unwrapped) && isEnumValueClass(gs, rhs_unwrapped)) {
```

elliottt replied:
What do you think about reordering this conditional to check if rhs is a subtype of lhs last?

elliottt replied:
What about rewriting this to check that RHS is a subtype of LHS, and LHS is a ClassType with a TEnum name? (You can use the `NameRef::isTEnumName` helper to test that.)

aprocter-stripe commented:
@elliottt Think I got everything, and also added a test for "subset" enums as you pointed out in Slack. (Those actually did seem to work with the old approach (checking for grandchildren of T::Enum), and they also work with the new (checking the name kind).)

elliottt reviewed (approved):
I think this looks great, thanks for all the tests!

aprocter-stripe commented on core/types/calls.cc:
```diff
@@ -3072,6 +3072,58 @@ class Module_tripleEq : public IntrinsicMethod {
+//  class TotallyUnrelatedThing ; end
+//
+// returns true for "X" and "Y", but false for "C", "T::Enum", and "TotallyUnrelatedThing".
+static bool isEnumValueClass(const GlobalState &gs, const TypePtr &type) {
+    bool must_exist = false;
+    auto unwrapped = unwrapSymbol(gs, type, must_exist);
+
+    if (!unwrapped.exists()) {
+        return false;
+    }
+
+    return unwrapped.data(gs)->name.isTEnumName(gs);
+}
```
Good call --- I think inlining is the way to go. (Started writing it this way 'cause I thought I'd need it in more than one place.) Will do that before merge.

aprocter-stripe commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JqxlF7P4JZZNMR
→ https://go/builds/bui_JqxmI6BrAFpFVW

aprocter-stripe commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_JrJaYE9CYrk26d
→ https://go/builds/bui_JrJaP2T0kFNkSY
